lib/temporal/activity/context.rb

```diff
@@ -1,21 +1,57 @@
+require 'temporal/error/failure'
+
 module Temporal
   class Activity
     class Context
       attr_reader :info
 
-      def initialize(info, heartbeat_proc)
+      def initialize(info, heartbeat_proc, shielded: false)
         @thread = Thread.current
         @info = info
         @heartbeat_proc = heartbeat_proc
+        @cancelled = false
+        @shielded = shielded
+        @mutex = Mutex.new
       end
 
       def heartbeat(*details)
         heartbeat_proc.call(*details)
       end
 
+      def shield(&block)
+        # The whole activity is shielded or called from a nested shield
+        return block.call if @shielded
+
+        mutex.synchronize do
+          @shielded = true
+          result = block.call
+          raise Temporal::Error::CancelledError, 'Unhandled cancellation' if @cancelled
+
+          result
+        ensure # runs while still holding the lock
+          @shielded = false
+        end
+      end
+
+      def cancelled?
+        @cancelled
+      end
+
+      def cancel
```

Comment:
Any way to make this non-public API?

Reply:
I can make it private and do a cheeky `context.send(:cancel)`, but that's bad practise. Overall I don't see a reason to hide `cancel` from users (not that they need it), but they can't do much harm with it

Reply:
I also think this is a smell to be exposed as public method specially if the `cancel` method on the reactor thread gets more complex. I think it would make sense to go with `send(:cacel)` for now if we want to avoid cross thread communication at this point.

Reply:
It doesn't need to be called from a reactor thread, it's perfectly safe to call from an activity thread. It would be analogous to raising a `CancelledError`. You can also call it from a `shield` and it will make the shield raise when finished. So as I said — no problem if someone decides to call it from an activity.

Reply:
@antstorm  So if I understand this correctly a running activity can call cancel to cancel itself? Doesn;t this go against the temporal design where the cancel should come as a command instead? I'm trying to understand what the use cases should be here?

Reply:
> Doesn;t this go against the temporal design where the cancel should come as a command from the server instead

It suspect won't actually appear as a cancel from a Temporal POV. In fact, @antstorm - I suspect if a user throws a cancelled error outside of it being cancelled via core, core may fail on you if you send it an activity-cancelled. An activity can't cancel itself. I think where you have `if failure.canceled_failure_info` you should do a `if cancel_requested`.

On this specific point, a user can raise any exception they want, including our exceptions. We'd just treat a self-cancel as an activity failure like any other exception I suspect (once the change I suggested is made).

Reply:
Ah, very good point, I'll make the change 👍

Reply:
https://github.com/temporalio/sdk-ruby/pull/90/commits/5e81346c4ef0573a2590d6ecba8b6977a653265d